### DISCUSSION TITLE
Namespace conflict for type Channels.Message in Callback contract between Client and Server

### DISCUSSION THREAD

**Discussion by Samba143** [CoreWCF/CoreWCF#1444] [Q&A]
I would like to define service callback contract as below.
public interface IMyServiceCallback
{
      void ReportProgress(Message msg);
}
In traditional WCF, the type Message is coming from System.ServiceModel.Channels and no issue for service implementation and client to consume it.

In CoreWCF, the type Message is available in both namespaces CoreWCF.Channels and System.ServiceModel.Channels.
My doubt is what namespace I should use in service contract file? Below are few challenges I encountered.
1. If namespace System.ServiceModel.Channels is used, I have conflicts in service implementation class  with namespace CoreWCF for Message and other types.
2. If CoreWCF.Channels namespace is used, client will not be able to implement this callback (we shouldn't use CoreWCF packages on client)

One workaround is create two variants of contact assemblies one for client (using System.ServiceModel.Channels) and one for CoreWCF service (using CoreWCF.Channels). I would like to avoid this repetition if there is any better solution to this problem.

Is there any guideline for the usage of System.ServiceModel in CoreWCF server side code?
Appreciate your views to the above problem.

**Comment by Samba143** [comment]
@mconnew, 
Could you please share your thoughts for the above issue?

**Reply by mconnew** [reply]
While we have some runtime compatibility code to be able to convert System.ServiceModel contracts to CoreWCF contracts, it's only possible for object model style types. E.g. System.ServiceModel.ServiceContract doesn't hold any data other than the well defined public properties such as Namespace. Unfortunately there are a handful of scenarios where we can't provide automatic compatibility. One of those is when you use the Message class as a return type or parameter. In those cases you can't do code sharing between client and server projects without using `#if` conditional code blocks to switch the namespace. The communication will still work fine with the client ChannelFactory using System.ServiceModel.Channels.Message in the client contract interface and the CoreWCF service using CoreWCF.Channels.Message in the service contract interface, but you can't share the exact same runtime type between them. This also means you can't use a common assembly defining the contract that's shared between client and server.  
If you must be able to share the same interface, there might be a way to achieve what you need using a combination of MessageContractAttribute and OperationContextScope. If this is the case, can you explain your use case a bit more and I might be able to suggest alternatives.
